### PR TITLE
[SPARK-16074] [MLLIB] expose VectorUDT/MatrixUDT in a public API

### PR DESCRIPTION
**Pull request by mengxr**
## What changes were proposed in this pull request?

Both VectorUDT and MatrixUDT are private APIs, because UserDefinedType itself is private in Spark. However, in order to let developers implement their own transformers and estimators, we should expose both types in a public API to simply the implementation of transformSchema, transform, etc. Otherwise, they need to get the data types using reflection.
## How was this patch tested?

Unit tests in Scala and Java.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/linalg/dataTypes.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * :: DeveloperApi ::
+ * SQL data types for vectors and matrices.
+ */
+@DeveloperApi
+object sqlDataTypes {
+
+  /** Data type for [[Vector]]. */
+  lazy val VectorType: DataType = new VectorUDT
```
Comment (thunterdb) [inline mllib/src/main/scala/org/apache/spark/ml/linalg/dataTypes.scala]: why is it `lazy`?

Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/linalg/dataTypes.scala]: Does it hurt?

**Comment by liancheng** [comment]
LGTM

**Comment by liancheng** [comment]
Merging to master and branch-2.0 on behalf of @mengxr.

**Comment by rxin** [comment]
I've merged this in master/2.0.

**Comment by liancheng** [comment]
@rxin Thanks, failed to merge it due to network issue...

**Pull request closed without merging**